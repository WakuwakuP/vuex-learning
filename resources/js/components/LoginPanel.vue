<template>
  <div class="panel">
    <form class="form" @submit.prevent="login">
      <div v-if="loginErrors" class="errors">
        <ul v-if="loginErrors.name">
          <li v-for="msg in loginErrors.name" :key="msg">{{ msg }}</li>
        </ul>
        <ul v-if="loginErrors.email">
          <li v-for="msg in loginErrors.email" :key="msg">{{ msg }}</li>
        </ul>
        <ul v-if="loginErrors.password">
          <li v-for="msg in loginErrors.password" :key="msg">{{ msg }}</li>
        </ul>
      </div>
      <label for="login-email">Email</label>
      <input type="text" class="form__item" id="login-email" v-model="loginForm.email">
      <label for="login-password">Password</label>
      <input type="password" class="form__item" id="login-password" v-model="loginForm.password">
      <div class="form__button">
        <button type="submit" class="button button--inverse" id="login-submit">login</button>
      </div>
    </form>
  </div>
</template>

<script>
import { mapState } from "vuex";

export default {
  name: "LoginPanel",
  data() {
    return {
      loginForm: {
        email: "",
        password: ""
      }
    };
  },
  methods: {
    async login() {
      // authストアのloginアクションを呼び出す
      await this.$store.dispatch("auth/login", this.loginForm);

      // トップページに移動する
      if (this.apiStatus) {
        // トップページに移動する
        this.$router.push("/");
      }
    }
  },
  computed: {
    ...mapState({
      apiStatus: state => state.auth.apiStatus,
      loginErrors: state => state.auth.loginErrorMessages
    })
  }
};
</script>
