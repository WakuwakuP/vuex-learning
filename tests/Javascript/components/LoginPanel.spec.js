import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';

import LoginPanel from '@/components/LoginPanel';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('LoginPanel', () => {
  let actions;
  let options;
  let $router;

  // ストアのモック作成
  beforeEach(() => {
    // pushをモック化しないとエラーが出る
    $router = {
      push: jest.fn(),
    };
    actions = {
      'auth/login': jest.fn(),
    };
    const store = new Vuex.Store({
      state: {
        auth: {
          apiStatus: true,
          loginErrors: null,
        },
      },
      actions,
    });
    options = {
      store,
      localVue,
      // $routerをモック化
      mocks: {
        $router,
      },
    };
  });

  it('is a Vue Instance', () => {
    const wrapper = shallowMount(LoginPanel, options);
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  it('renders correctly', () => {
    const wrapper = shallowMount(LoginPanel, options);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('login Action', () => {
    const email = 'example@example.com';
    const password = 'jglr;akdjvnre:,f';
    const wrapper = shallowMount(LoginPanel, options);
    wrapper.find('#login-email').setValue(email);
    wrapper.find('#login-password').setValue(password);
    wrapper.find('form').trigger('submit.prevent');
    expect(actions['auth/login'])
      .toHaveBeenCalledWith(
        expect.anything(),
        { email, password },
        undefined,
      );
  });

  it('apiStatus is false', () => {
    const push = jest.fn();
    const store = new Vuex.Store({
      state: {
        auth: {
          apiStatus: false,
          loginErrors: null,
        },
      },
      actions: {
        'auth/login': jest.fn(),
      },
    });
    const mocks = {
      $$router: {
        push: jest.fn(),
      },
    };
    const email = 'example@example.com';
    const password = 'jglr;akdjvnre:,f';
    const wrapper = shallowMount(LoginPanel, {
      store,
      localVue,
      mocks,
    });
    wrapper.find('#login-email').setValue(email);
    wrapper.find('#login-password').setValue(password);
    wrapper.find('form').trigger('submit.prevent');
    expect(push).not.toBeCalled();
  });
});
