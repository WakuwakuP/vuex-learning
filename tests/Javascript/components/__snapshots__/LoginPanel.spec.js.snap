// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPanel renders correctly 1`] = `
<div
  class="panel"
>
  <form
    class="form"
  >
    <!---->
     
    <label
      for="login-email"
    >
      Email
    </label>
     
    <input
      class="form__item"
      id="login-email"
      type="text"
    />
     
    <label
      for="login-password"
    >
      Password
    </label>
     
    <input
      class="form__item"
      id="login-password"
      type="password"
    />
     
    <div
      class="form__button"
    >
      <button
        class="button button--inverse"
        type="submit"
      >
        login
      </button>
    </div>
  </form>
</div>
`;
